File: Dot examples/ASP Book Examples/Book Examples/Chapter19/SqlHierarchicalDataSourceControlVB/App_Code/SqlHierarchicalDataSourceView.vb
Imports System
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Data
Imports System.Data.SqlClient

Namespace AspNetUnleashed

    ''' <summary>
    ''' Represents the data returned from database
    ''' </summary>
    Public Class SqlHierarchicalDataSourceView
        Inherits HierarchicalDataSourceView

        Private _owner As SqlHierarchicalDataSource
        Private _viewPath As String
        Private _data As DataTable

        Public Sub New(ByVal owner As SqlHierarchicalDataSource, ByVal viewPath As String)
            _owner = owner
            _viewPath = viewPath
        End Sub

        ''' <summary>
        ''' The DataTable which contains all rows from
        ''' underlying database table
        ''' </summary>
        Public ReadOnly Property Data() As DataTable
            Get
                Return _data
            End Get
        End Property

        ''' <summary>
        ''' We need to expose this for the SqlNodes
        ''' </summary>
        Public ReadOnly Property DataKeyName() As String
            Get
                Return _owner.DataKeyName
            End Get
        End Property

        ''' <summary>
        ''' We need to expose this for the SqlNodes
        ''' </summary>
        Public ReadOnly Property DataParentKeyName() As String
            Get
                Return _owner.DataParentKeyName
            End Get
        End Property

        ''' <summary>
        ''' Get the top-level rows (rows without parent rows)
        ''' </summary>
        Public Overrides Function [Select]() As IHierarchicalEnumerable
            ' Verify DataKeyName and DataParentKeyName properties
            If String.IsNullOrEmpty(DataKeyName) Then
                Throw New Exception("You must set the DataKeyName property")
            End If
            If String.IsNullOrEmpty(DataParentKeyName) Then
                Throw New Exception("You must set the DataParentKeyName property")
            End If

            ' Return DataView from SqlDataSource 
            If _owner.DataSourceMode <> SqlDataSourceMode.DataSet Then
                Throw New Exception("DataSourceMode must be set to DataSet")
            End If
            Dim view As DataView = CType(_owner.Select(DataSourceSelectArguments.Empty), DataView)
            _data = view.Table

            ' Get the root rows
            Dim filter As String = String.Format("{0} IS NULL", Me.DataParentKeyName)
            Dim rootRows() As DataRow = _data.Select(filter)

            ' Build up the hierarchical collection
            Dim en As SqlHierarchicalEnumerable = New SqlHierarchicalEnumerable()
            For Each row As DataRow In rootRows
                en.Add(New SqlNode(Me, row))
            Next
            Return en
        End Function

    End Class
End Namespace

File: Dot examples/ASP Book Examples/Book Examples/Chapter19/SqlHierarchicalDataSourceControlCS/App_Code/SqlHierarchicalDataSourceView.cs
using System;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Data;
using System.Data.SqlClient;

namespace AspNetUnleashed
{

    /// <summary>
    /// Represents the data returned from database
    /// </summary>
    public class SqlHierarchicalDataSourceView : HierarchicalDataSourceView
    {
        private SqlHierarchicalDataSource _owner;
        private string _viewPath;
        private DataTable _data;

        public SqlHierarchicalDataSourceView(SqlHierarchicalDataSource owner, string viewPath)
        {
            _owner = owner;
            _viewPath = viewPath;
        }

        /// <summary>
        /// The DataTable which contains all rows from
        /// underlying database table
        /// </summary>
        public DataTable Data
        {
            get { return _data; }
        }

        /// <summary>
        /// We need to expose this for the SqlNodes
        /// </summary>
        public string DataKeyName
        {
            get { return _owner.DataKeyName; }
        }

        /// <summary>
        /// We need to expose this for the SqlNodes
        /// </summary>
        public string DataParentKeyName
        {
            get { return _owner.DataParentKeyName; }
        }

        /// <summary>
        /// Get the top-level rows (rows without parent rows)
        /// </summary>
        /// <returns></returns>
        public override IHierarchicalEnumerable Select()
        {
            // Verify DataKeyName and DataParentKeyName properties
            if (String.IsNullOrEmpty(DataKeyName))
                throw new Exception("You must set the DataKeyName property");
            if (String.IsNullOrEmpty(DataParentKeyName))
                throw new Exception("You must set the DataParentKeyName property");

            // Return DataView from SqlDataSource 
            if (_owner.DataSourceMode != SqlDataSourceMode.DataSet)
                throw new Exception("DataSourceMode must be set to DataSet");
            DataView view = (DataView)_owner.Select(DataSourceSelectArguments.Empty);
            _data = view.Table;

            // Get the root rows
            string filter = string.Format("{0} IS NULL", this.DataParentKeyName);
            DataRow[] rootRows = _data.Select(filter);

            // Build up the hierarchical collection
            SqlHierarchicalEnumerable en = new SqlHierarchicalEnumerable();
            foreach (DataRow row in rootRows)
                en.Add(new SqlNode(this, row));
            return en;
        }


    }
}